File: scripts/benchmark-detector.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	gozstd "github.com/DataDog/zstd"
	"github.com/goccy/go-json"
)

// Minimal operation struct
type Operation struct {
	DID string `json:"did"`
	//Operation map[string]interface{} `json:"operation"`
	CID       string      `json:"cid"`
	Nullified interface{} `json:"nullified,omitempty"`
	CreatedAt time.Time   `json:"createdAt"`

	// RawJSON stores the original JSON bytes for exact reproduction
	RawJSON []byte `json:"-"`
}

// User's detect function
func detect(op *Operation) []string {
	labels := []string{}

	if strings.HasPrefix(op.DID, "did:plc:aa") {
		labels = append(labels, "test")
	}

	return labels
}

func main() {
	bundleDir := "./"
	if len(os.Args) > 1 {
		bundleDir = os.Args[1]
	}

	startBundle := 1
	if len(os.Args) > 2 {
		startBundle, _ = strconv.Atoi(os.Args[2])
	}

	endBundle := 100
	if len(os.Args) > 3 {
		endBundle, _ = strconv.Atoi(os.Args[3])
	}

	fmt.Fprintf(os.Stderr, "Processing bundles %d-%d from %s\n", startBundle, endBundle, bundleDir)
	fmt.Fprintf(os.Stderr, "\n")

	// Buffered stdout
	writer := bufio.NewWriterSize(os.Stdout, 512*1024)
	defer writer.Flush()

	// CSV header
	writer.WriteString("bundle,position,cid,size,confidence,labels\n")

	totalOps := 0
	matchCount := 0
	totalBytes := int64(0)
	matchedBytes := int64(0)

	startTime := time.Now()

	for bundleNum := startBundle; bundleNum <= endBundle; bundleNum++ {
		bundleFile := filepath.Join(bundleDir, fmt.Sprintf("%06d.jsonl.zst", bundleNum))

		// Read compressed file
		compressed, err := os.ReadFile(bundleFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "\nError reading bundle %d: %v\n", bundleNum, err)
			continue
		}

		// Decompress with native gozstd (CGO, fast as native zstd)
		decompressed, err := gozstd.Decompress(nil, compressed)
		if err != nil {
			fmt.Fprintf(os.Stderr, "\nError decompressing bundle %d: %v\n", bundleNum, err)
			continue
		}

		// Split into lines
		lines := bytes.Split(decompressed, []byte("\n"))

		for position, line := range lines {
			if len(line) == 0 {
				continue
			}

			totalOps++
			opSize := len(line)
			totalBytes += int64(opSize)

			// Parse JSON with go-json
			var op Operation
			if err := json.Unmarshal(line, &op); err != nil {
				fmt.Fprintf(os.Stderr, "Error parsing operation: %v\n", err)
				continue
			}

			// Run detect function
			labels := detect(&op)

			if len(labels) > 0 {
				matchCount++
				matchedBytes += int64(opSize)

				// Extract last 4 chars of CID
				cidShort := op.CID
				if len(cidShort) > 4 {
					cidShort = cidShort[len(cidShort)-4:]
				}

				// Write CSV line
				writer.WriteString(strconv.Itoa(bundleNum))
				writer.WriteByte(',')
				writer.WriteString(strconv.Itoa(position))
				writer.WriteByte(',')
				writer.WriteString(cidShort)
				writer.WriteByte(',')
				writer.WriteString(strconv.Itoa(opSize))
				writer.WriteByte(',')
				writer.WriteString("0.95")
				writer.WriteByte(',')
				writer.WriteString(strings.Join(labels, ";"))
				writer.WriteByte('\n')
			}
		}

		// Progress
		if bundleNum%10 == 0 {
			elapsed := time.Since(startTime).Seconds()
			opsPerSec := float64(totalOps) / elapsed
			fmt.Fprintf(os.Stderr, "Processed %d/%d bundles | %d ops | %.0f ops/sec\r",
				bundleNum, endBundle, totalOps, opsPerSec)
		}
	}

	elapsed := time.Since(startTime).Seconds()

	// Stats
	fmt.Fprintf(os.Stderr, "\n\n")
	fmt.Fprintf(os.Stderr, "✓ Detection complete\n")
	fmt.Fprintf(os.Stderr, "  Total operations:   %d\n", totalOps)
	fmt.Fprintf(os.Stderr, "  Matches found:      %d (%.2f%%)\n", matchCount, float64(matchCount)/float64(totalOps)*100)
	fmt.Fprintf(os.Stderr, "  Total size:         %.1f MB\n", float64(totalBytes)/1e6)
	fmt.Fprintf(os.Stderr, "  Matched size:       %.1f MB (%.2f%%)\n", float64(matchedBytes)/1e6, float64(matchedBytes)/float64(totalBytes)*100)
	fmt.Fprintf(os.Stderr, "\n")
	fmt.Fprintf(os.Stderr, "  Time elapsed:       %.2fs\n", elapsed)
	fmt.Fprintf(os.Stderr, "  Throughput:         %.0f ops/sec\n", float64(totalOps)/elapsed)
	fmt.Fprintf(os.Stderr, "  Speed:              %.1f MB/sec\n", float64(totalBytes)/elapsed/1e6)
}

File: scripts/benchmark-detector.js
#!/usr/bin/env bun

// User's detect function
function detect({ op }) {
  const labels = [];
  
  if (op.did.startsWith('did:plc:aa')) {
    labels.push('test')
  }

  console.log(op.operation.sig)
  
  return labels;
}

// ==========================================
// Pure Bun bundle processor with native zstd
// ==========================================

const BUNDLE_DIR = process.argv[2] || './';
const START_BUNDLE = parseInt(process.argv[3]) || 1;
const END_BUNDLE = parseInt(process.argv[4]) || 100;

console.error(`Processing bundles ${START_BUNDLE}-${END_BUNDLE} from ${BUNDLE_DIR}`);
console.error('');

// CSV header
console.log('bundle,position,cid,size,confidence,labels');

let totalOps = 0;
let matchCount = 0;
let totalBytes = 0;
let matchedBytes = 0;

const startTime = Date.now();

for (let bundleNum = START_BUNDLE; bundleNum <= END_BUNDLE; bundleNum++) {
  const bundleFile = `${BUNDLE_DIR}/${bundleNum.toString().padStart(6, '0')}.jsonl.zst`;
  
  try {
    // Read compressed bundle
    const compressed = await Bun.file(bundleFile).arrayBuffer();
    
    // Decompress using native Bun zstd (FAST!)
    const decompressed = Bun.zstdDecompressSync(compressed);
    
    // Convert to text
    const text = new TextDecoder().decode(decompressed);
    
    const lines = text.split('\n').filter(line => line.trim());
    
    for (let position = 0; position < lines.length; position++) {
      const line = lines[position];
      if (!line) continue;
      
      totalOps++;
      const opSize = line.length;
      totalBytes += opSize;
      
      try {
        const op = JSON.parse(line);
        const labels = detect({ op });
        
        if (labels && labels.length > 0) {
          matchCount++;
          matchedBytes += opSize;
          
          // Extract last 4 chars of CID
          const cidShort = op.cid.slice(-4);
          
          console.log(
            `${bundleNum},${position},${cidShort},${opSize},0.95,${labels.join(';')}`
          );
        }
      } catch (err) {
        console.error(`Error parsing operation: ${err.message}`);
      }
    }
    
    // Progress
    if (bundleNum % 10 === 0) {
      const elapsed = (Date.now() - startTime) / 1000;
      const opsPerSec = (totalOps / elapsed).toFixed(0);
      console.error(`Processed ${bundleNum}/${END_BUNDLE} bundles | ${totalOps} ops | ${opsPerSec} ops/sec\r`);
    }
    
  } catch (err) {
    console.error(`\nError loading bundle ${bundleNum}: ${err.message}`);
  }
}

const elapsed = (Date.now() - startTime) / 1000;

// Stats
console.error('\n');
console.error('✓ Detection complete');
console.error(`  Total operations:   ${totalOps}`);
console.error(`  Matches found:      ${matchCount} (${(matchCount/totalOps*100).toFixed(2)}%)`);
console.error(`  Total size:         ${(totalBytes / 1e6).toFixed(1)} MB`);
console.error(`  Matched size:       ${(matchedBytes / 1e6).toFixed(1)} MB (${(matchedBytes/totalBytes*100).toFixed(2)}%)`);
console.error('');
console.error(`  Time elapsed:       ${elapsed.toFixed(2)}s`);
console.error(`  Throughput:         ${(totalOps / elapsed).toFixed(0)} ops/sec`);
console.error(`  Speed:              ${(totalBytes / elapsed / 1e6).toFixed(1)} MB/sec`);
